From: Stanislav Malyshev <stas@php.net>
Date: Sat, 2 Mar 2019 15:07:40 -0800
Subject: Fix bug #77563 - Uninitialized read in exif_process_IFD_in_MAKERNOTE

Also fix for bug #77659

(cherry picked from commit 8ac6fee8562533a15db90062117210ed28b44fea)
---
 ext/exif/exif.c              |   3 ++-
 ext/exif/tests/bug77563.jpg  | Bin 0 -> 63 bytes
 ext/exif/exif.c              |    3 ++-
 ext/exif/tests/bug77563.jpg  |binary
 ext/exif/exif.c              |    3 ++-
 ext/exif/tests/bug77563.phpt |   16 ++++++++++++++++
 2 files changed, 18 insertions(+), 1 deletion(-)
 create mode 100644 ext/exif/tests/bug77563.jpg
 create mode 100644 ext/exif/tests/bug77563.phpt

--- a/ext/exif/exif.c
+++ b/ext/exif/exif.c
@@ -2751,7 +2751,7 @@ static int exif_process_IFD_in_MAKERNOTE
 		break;
 	}
 
-	if (maker_note->offset >= value_len) {
+	if (value_len < 2 || maker_note->offset >= value_len - 1) {
 		/* Do not go past the value end */
 		exif_error_docref("exif_read_data#error_ifd" EXIFERR_CC, ImageInfo, E_WARNING, "IFD data too short: 0x%04X offset 0x%04X", value_len, maker_note->offset);
 		return FALSE;
@@ -2804,6 +2804,7 @@ static int exif_process_IFD_in_MAKERNOTE
 			break;
 		default:
 		case MN_OFFSET_NORMAL:
+			data_len = value_len;
 			break;
 	}
 
--- /dev/null
+++ b/ext/exif/tests/bug77563.phpt
@@ -0,0 +1,16 @@
+--TEST--
+Bug 77563 (Uninitialized read in exif_process_IFD_in_MAKERNOTE)
+--SKIPIF--
+<?php if (!extension_loaded('exif')) print 'skip exif extension not available';?>
+--FILE--
+<?php
+$s = exif_thumbnail(__DIR__."/bug77563.jpg");
+?>
+DONE
+--EXPECTF--
+Warning: exif_thumbnail(bug77563.jpg): Illegal IFD offset in %s/bug77563.php on line %d
+
+Warning: exif_thumbnail(bug77563.jpg): File structure corrupted in %s/bug77563.php on line %d
+
+Warning: exif_thumbnail(bug77563.jpg): Invalid JPEG file in %s/bug77563.php on line %d
+DONE
\ No newline at end of file
